File: bbbyGitHub/core/src/main/java/com/bbby/aem/core/util/ImageRendition.java
package com.bbby.aem.core.util;

import org.apache.commons.lang3.StringUtils;

/**
 * Canonical record of all web/JPEG renditions supported by this site.
 * <p>
 * The JavaScript version of this file is
 * {@code /apps/bbby/components/content/image/imageRendition.js}
 * <p>
 * Any changes to the Java or JavaScript version of this file should be propagated to both.
 *
 * @author joelepps
 *
 */
public enum ImageRendition {

    LANDSCAPE_X2(3600, 3600),
    LANDSCAPE_X1(1920, 1920),
    PORTRAIT_X2(1200, 1200),
    PORTRAIT_X1(600, 600),

    ORIGINAL(0, 0);

    private final int width;
    private final int height;

    ImageRendition(int width, int height) {
        this.width = width;
        this.height = height;
    }

    public String getRendition(String imagePath) {
        if (StringUtils.isBlank(imagePath)) return imagePath;
        if (imagePath.contains("://")) return imagePath;

        if (this == ORIGINAL) {
            String ext = getExtension(imagePath);
            return imagePath + ".imgo." + ext;
        }
        // web images should always be jpg extension
        return imagePath + ".imgw." + width + "." + height + "." + "jpg";
    }

    private String getExtension(String imagePath) {
        if (imagePath != null && imagePath.contains(".png")) {
            return "png";
        } else if (imagePath != null && imagePath.contains(".jpg")) {
            return "jpg";
        } else if (imagePath != null && imagePath.contains(".svg")) {
            return "svg";
        } else {
            return "jpg"; // default
        }
    }

    public int getWidth() {
        return width;
    }

    public int getHeight() {
        return height;
    }

}

File: bbbyGitHub/ui.apps/src/main/gulp/common/js/utils/imageRendition.js
'use strict';

/**
 * Module is the JS equivalent of com.bbby.aem.core.util.ImageRendition
 *
 * Provides utility methods to convert an image path to one with rendition selectors.
 *
 * Any changes to the Java or JavaScript version of this file should be propagated to both.
 */

export default {
    LANDSCAPE_X2: new Rendition(3600, 3600),
    LANDSCAPE_X1: new Rendition(1920, 1920),
    PORTRAIT_X2: new Rendition(1200, 1200),
    PORTRAIT_X1: new Rendition(600, 600),

    ORIGINAL: new Rendition(0, 0)
}

function Rendition(width, height) {
    this.width = width;
    this.height = height;
}

Rendition.prototype.getRendition = function(imagePath) {
    if (!imagePath) return imagePath;

    if (this.width == 0 || this.height == 0) {
        var ext = getExtension(imagePath);
        return imagePath + '.imgo.' + ext;
    }

    // web images should always be jpg extension
    return imagePath + ".imgw." + this.width + "." + this.height + "." + "jpg";
};

function getExtension(imagePath) {
    if (imagePath && imagePath.indexOf(".png") > 0) {
        return "png";
    } else if (imagePath && imagePath.indexOf(".jpg") > 0) {
        return "jpg";
    } else if (imagePath && imagePath.indexOf(".svg") > 0) {
        return "svg";
    } else {
        return "jpg"; // default
    }
}
